pkg/buildpipeline/pipeline.go

```diff
@@ -772,79 +808,123 @@ func createPipelineTasks(block *TransformedStage, blocks []TransformedStage, hea
-			Inputs: []pipelinev1alpha1.PipelineTaskInputResource{{
-				Name:     "workspace",
-				Resource: "common-workspace",
-				From:     provider,
-			}},
+			Inputs: []pipelinev1alpha1.PipelineTaskInputResource{
+				{
+					Name:     "workspace",
+					Resource: "common-workspace",
+					From:     provider,
+				},
+				{
+					// TODO: Switch from this kind of hackish approach to non-resource-based dependencies once they land.
+					Name:     "temp-ordering-resource",
+					Resource: "temp-ordering-resource",
+					From:     previousStageNames,
+				},
+			},
 		}
-
 		block.PipelineTask = &pTask
-		*head = append(*head, *block)
 
 		return []pipelinev1alpha1.PipelineTask{pTask}
 	}
 }
 
-// Looks for the most recent Task using the desired workspace that was not in a parallel block and returns whether such a workspace was found and the name of the corresponding task
-func findWorkspaceProvider(block TransformedStage, blocks []TransformedStage) (bool, []string) {
-	if *block.Stage.Options.Workspace == "empty" {
+// Looks for the most recent Task using the desired workspace that was not in the
+// same parallel block and returns the name of the corresponding Task.
+func findWorkspaceProvider(stage, sibling *TransformedStage) (bool, []string) {
+	if *stage.Stage.Options.Workspace == "empty" {
 		return true, nil
 	}
 
-	var parallelEndTasks []string
-
-	for i := len(blocks) - 1; i >= 0; i-- {
-		previousBlock := blocks[i]
-		if len(previousBlock.Sequential) > 0 {
-			found, provider := findWorkspaceProvider(block, previousBlock.Sequential)
+	for sibling != nil {
+		if sibling.isSequentialStage() {
+			found, provider := findWorkspaceProvider(stage, sibling.Sequential[len(sibling.Sequential)-1])
 			if found {
-				return true, append(parallelEndTasks, provider...)
+				return true, provider
+			}
+		} else if sibling.isParallelStage() {
+			// We don't want to use a workspace from a parallel block outside of that block,
+			// but we do need to descend inwards in case stage is in that same block.
+			if stage.getEnclosing(sibling.Depth) == sibling {
+				for _, nested := range sibling.Parallel {
+					// Pick the parallel branch that has stage
+					if stage.getEnclosing(nested.Depth) == nested {
+						found, provider := findWorkspaceProvider(stage, nested)
+						if found {
+							return true, provider
+						}
+					}
+				}
 			}
-		} else if len(previousBlock.Parallel) > 0 {
-			// We don't want to use a workspace from a parallel block outside of that block.
 			// TODO: What to do about custom workspaces? Check for erroneous uses specially?
 			// Allow them if only one of the parallel tasks uses the same resource?
-
-			if previousBlock.Depth == block.Depth {
-				// TODO: We do want to record the last task in each parallel branch, though, currently hacked in as from,
-				// but will end up being used for the non-resource-based dependency ordering field.
-				parallelEndTasks = findEndTasks(previousBlock)
```

Comment:
Now this stuff is handled with `temp-ordering-resource` using `findPreviousNonBlockStages`